src/lib/resolve_type.hpp

```diff
@@ -169,78 +169,78 @@ void resolve_data_type(DataType data_type, const Functor& func) {
+/*void*/ resolve_segment_type(BaseSegmentType& segment, const Functor& func) {
+  using ValueSegmentPtr = ConstOutIfConstIn<BaseSegmentType, ValueSegment<CxlumnDataType>>*;
+  using ReferenceSegmentPtr = ConstOutIfConstIn<BaseSegmentType, ReferenceSegment>*;
+  using EncodedSegmentPtr = ConstOutIfConstIn<BaseSegmentType, BaseEncodedSegment>*;
 
-  if (auto value_column = dynamic_cast<ValueColumnPtr>(&column)) {
-    func(*value_column);
-  } else if (auto ref_column = dynamic_cast<ReferenceColumnPtr>(&column)) {
-    func(*ref_column);
-  } else if (auto encoded_column = dynamic_cast<EncodedColumnPtr>(&column)) {
-    resolve_encoded_column_type<ColumnDataType>(*encoded_column, func);
+  if (auto value_segment = dynamic_cast<ValueSegmentPtr>(&segment)) {
+    func(*value_segment);
+  } else if (auto ref_segment = dynamic_cast<ReferenceSegmentPtr>(&segment)) {
+    func(*ref_segment);
+  } else if (auto encoded_segment = dynamic_cast<EncodedSegmentPtr>(&segment)) {
+    resolve_encoded_segment_type<CxlumnDataType>(*encoded_segment, func);
   } else {
-    Fail("Unrecognized column type encountered.");
+    Fail("Unrecognized cxlumn type encountered.");
   }
 }
 
 /**
- * Resolves a data type by passing a hana::type object and the downcasted column on to a generic lambda
+ * Resolves a data type by passing a hana::type object and the downcasted segment on to a generic lambda
  *
- * @param data_type is an enum value of any of the supported column types
+ * @param data_type is an enum value of any of the supported cxlumn types
  * @param func is a generic lambda or similar accepting two parameters: a hana::type object and
- *   a reference to a specialized column (value, dictionary, reference)
+ *   a reference to a specialized segment (value, dictionary, reference)
  *
  *
  * Example:
  *
  *   template <typename T>
- *   void process_column(hana::basic_type<T> type, ValueColumn<T>& column);
+ *   void process_segment(hana::basic_type<T> type, ValueSegment<T>& segment);
  *
  *   template <typename T>
- *   void process_column(hana::basic_type<T> type, DictionaryColumn<T>& column);
+ *   void process_segment(hana::basic_type<T> type, DictionarySegment<T>& segment);
  *
  *   template <typename T>
- *   void process_column(hana::basic_type<T> type, ReferenceColumn& column);
+ *   void process_segment(hana::basic_type<T> type, ReferenceSegment& segment);
  *
- *   resolve_data_and_column_type(base_column, [&](auto type, auto& typed_column) {
- *     process_column(type, typed_column);
+ *   resolve_data_and_cxlumn_type(base_segment, [&](auto type, auto& typed_segment) {
+ *     process_segment(type, typed_segment);
  *   });
  */
-template <typename Functor, typename BaseColumnType>  // BaseColumnType allows column to be const and non-const
-std::enable_if_t<std::is_same<BaseColumn, std::remove_const_t<BaseColumnType>>::value>
-/*void*/ resolve_data_and_column_type(BaseColumnType& column, const Functor& func) {
```

Comment:
@mrks 
Is there a reason you renamed `resolve_column_type` to `resolve_segment_type` but not this one?
I find this quite confusing.

Reply:
Mmmh, no. Interestingly that did not come up in my grep

Reply:
Okay, so we will go with `resolve_data_and_segment_type`? I am currently in the process of making some changes, will include that.

Reply:
yes